tests/unit/test_charm.py

```diff
@@ -373,60 +404,73 @@ def test_install_apt_packages_add_package_failure(self, update, add_package):
+        repo_map.return_value = set()
+
         self.harness.charm._add_mongodb_org_repository()
-        get_mapping.return_value = {}
-        urlopen.assert_called()
         from_repo_line.assert_called()
+        (from_repo_line.return_value.import_key).assert_called_with(MONGODB_PUB_KEY)
+        self.assertEqual(repo_map.return_value, {from_repo_line.return_value})
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.from_repo_line")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.urlopen")
-    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(self, urlopen, get_mapping):
-        get_mapping.return_value = {}
+    def test_add_mongodb_org_repository_gpg_fail_leads_to_blocked(
+        self, urlopen, import_key, from_repo_line, repo_map
+    ):
+        # preset values
+        repo_map.return_value = set()
         urlopen.side_effect = URLError("urlopen error")
-        with self.assertLogs("charm", "ERROR") as logs:
-            self.harness.charm._add_mongodb_org_repository()
-            self.assertIn("ERROR:charm:failed to get GPG key, reason:", "".join(logs.output))
+        self.harness.charm._add_mongodb_org_repository()
 
+        # verify we don't add repo when an exception occurs and that we enter blocked state
+        from_repo_line.assert_not_called()
+        import_key.assert_not_called()
+        self.assertEqual(repo_map.return_value, set())
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't install MongoDB"))
 
     @patch("charm.apt.RepositoryMapping")
+    @patch("charm.apt.DebianRepository.import_key")
     @patch("charm.apt.DebianRepository.from_repo_line")
     @patch("charm.urlopen")
     def test_add_mongodb_org_repository_cant_create_list_file_blocks(
-        self, urlopen, from_repo_line, get_mapping
+        self, urlopen, from_repo_line, import_key, repo_map
     ):
-        get_mapping.return_value = {}
+        repo_map.return_value = set()
         exceptions = [
             apt.InvalidSourceError("invalid source message"),
             ValueError("value message"),
         ]
-        log_messages = [
-            "ERROR:charm:failed to add repository, invalid source: invalid source message",
-            "ERROR:charm:failed to add repository: value message",
-        ]
-        for exception, log_message in zip(exceptions, log_messages):
-            with self.assertLogs("charm", "ERROR") as logs:
-                from_repo_line.side_effect = exception
-                self.harness.charm._add_mongodb_org_repository()
-                self.assertIn(log_message, "".join(logs.output))
+        for exception in exceptions:
+            from_repo_line.side_effect = exception
+            self.harness.charm._add_mongodb_org_repository()
 
+            # verify we don't add repo when an exception occurs and that we enter blocked state
+            import_key.assert_not_called()
```

Comment:
This and the following line are not needed for the same reason as mentioned before.  In fact, logically they could not be called because they need the output from from_repo_line and we have set up the test so that call fails.

Reply:
I added the checks 
```
import_key.assert_not_called()
self.assertEqual(repo_map.return_value, set())
```
Because I in future implementations someone could restructure the code and we would want to be sure that if they had code failing from the call to `from_repo_line` that they their implementation still obeys the principle logic of "if `from_repo_line` fails do not `import_key` or `add` the repo.